crates/assists/src/handlers/infer_function_return_type.rs

```diff
@@ -0,0 +1,237 @@
+use hir::HirDisplay;
+use syntax::{ast, AstNode, SyntaxToken, TextSize};
+use test_utils::mark;
+
+use crate::{AssistContext, AssistId, AssistKind, Assists};
+
+// Assist: infer_function_return_type
+//
+// Adds the return type to a function or closure inferred from its tail expression if it doesn't have a return
+// type specified.
+//
+// ```
+// fn foo() { 4<|>2i32 }
+// ```
+// ->
+// ```
+// fn foo() -> i32 { 42i32 }
+// ```
+pub(crate) fn infer_function_return_type(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let expr = ctx.find_node_at_offset::<ast::Expr>()?;
+    let (tail_expr, insert_pos, wrap_expr) = extract_tail(expr)?;
+    let module = ctx.sema.scope(tail_expr.syntax()).module()?;
+    let ty = ctx.sema.type_of_expr(&tail_expr).filter(|ty| !ty.is_unit())?;
```

Comment:
Super minor nit, but I'd write this as

```rust
let ty = ctx.sema.type_of_expr(&tail_expr)?;
if ty.is_unit() {
    return None;
}
```

Significantly more verbose, but reduces congnitive load a bit.